### PR TITLE
python training api bindings

### PR DESCRIPTION
**Pull request by AdamLouly**
**Description**: **Python API Bindings for on device training. **
**Motivation and Context**
- This PR contains api bindings so python users can perform a whole training loop.

### CODE REVIEW

File: orttraining/orttraining/python/training/onnxblock/engine/training_module.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+
+class TrainingModule:
+    """
+    Class for running Training.
+    This class is a wrapper of Module Class.
+    """
+
+    def __init__(
+        self, train_model_uri, ckpt_uri, eval_model_uri=None, env=None, session_options=None, providers=None, **kwargs
```
Comment (ashbhandare) [inline orttraining/orttraining/python/training/onnxblock/engine/training_module.py:11]: A single env instance is maintained within pybind, so the env arg should be dropped.

**Comment by ashbhandare** [comment]
Where will the checkpointing functions fit in in this design? This differs in the way the APIs are exposed in C and C# and might cause some confusion. Though the audience for the python bindings will mostly be data scientists wanting to experiment, there might be a transition step involved in moving from python bindings to using C/C# APIs in production. In that case, we want to make their life as easy as possible.

**Comment by pengwa** [comment]
I had a similar question with @ashbhandare , if users have python env, they can use ORTModule or ORTTrainer(not recommended) apis. What's the purpose of adding a new Python binding apis, which originally was designed for federated learning or on-device trainings.

**Comment by AdamLouly** [comment]
> Where will the checkpointing functions fit in in this design? This differs in the way the APIs are exposed in C and C# and might cause some confusion. Though the audience for the python bindings will mostly be data scientists wanting to experiment, there might be a transition step involved in moving from python bindings to using C/C# APIs in production. In that case, we want to make their life as easy as possible.

for checkpointing, I'm projecting it to be called in this way for the end user:

to save checkpoint while training:
save(model.state_dict(),"path")

to load checkpoint into a model:
model.load_state_dict("path")

same for optimizer.

File: orttraining/orttraining/python/training/onnxblock/engine/training_module.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        self._train_model_uri = train_model_uri
+        self._ckpt_uri = ckpt_uri
+        self._eval_model_uri = eval_model_uri
+        self._session_options = session_options
+        if providers is None:
+            self._providers = C.get_available_providers()
+
+        self._create_training_module()
+
+    def __call__(self, *input):
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/engine/training_module.py:34]: Can you add a docstring for this method? Is the expectation that the inputs and outputs are numpy ndarrays? Or are they ORTValues?

File: orttraining/orttraining/python/training/onnxblock/engine/training_module.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+        if self.training:
+            self._model.train_step(*input, self.fetches)
+        elif self.training == False:
+            self._model.eval_step(*input, self.fetches)
+
+        return self.fetches[0].numpy()
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/engine/training_module.py:41]: What is the idea behind having fetches as a member variable?

File: orttraining/orttraining/python/training/onnxblock/engine/training_module.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+    def _create_training_module(self):
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/engine/training_module.py:43]: This can be a utility standalone function in a separate _utils.py file. It does not need to be a member function.

File: orttraining/orttraining/python/training/onnxblock/engine/training_module.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+        Args:
+            mode (bool): whether to set training mode (``True``) or evaluation
+                            mode (``False``). Default: ``True``.
+
+        Returns:
+            Module: self
+        """
+        if not isinstance(mode, bool):
+            raise ValueError("training mode is expected to be boolean")
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/engine/training_module.py:69]: Since python is not a typed language, this is not required. You can use type hints to indicate the expected type.

File: orttraining/orttraining/python/training/onnxblock/engine/training_module.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        """
+        Resets the gradient of the parameters.
+        """
+        return self._model.reset_grad()
+
+    def get_model(self):
+        """
+        Returns the model to be passed to initialize the Optimizer.
+        """
+        return self._model
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/engine/training_module.py:92]: Is this method needed?

File: orttraining/orttraining/python/training/onnxblock/engine/training_module.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        """
+        return self._model.save_checkpoint(ckpt_uri)
+
+    def loss(self):
+        """
+        Returns the loss value.
+        """
+        if len(self.fetches) == 0:
+            raise ValueError("You should run a train step before calling this function.")
+        return self.fetches[0].numpy()
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/engine/training_module.py:112]: We cannot assume that the first output of the model is a loss. This function is probably not needed.

File: orttraining/orttraining/python/training/onnxblock/engine/training_module.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+class TrainingModule:
+    """
+    Class for running Training.
+    This class is a wrapper of Module Class.
+    """
+
+    training: bool
+
+    def __init__(
+        self, train_model_uri, ckpt_uri, eval_model_uri=None, session_options=None, providers=None, **kwargs
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/engine/training_module.py:16]: Let's provide suitable defaults for session options and providers and not expose the in the constructor. We also don't need kwargs right now.

File: orttraining/orttraining/python/training/onnxblock/engine/training_module.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        """
+        # TODO : Add support for bytes on train_model_uri and eval_model_uri.
+        self.training = True
+        self.fetches = OrtValueVector()
+        self._train_model_uri = train_model_uri
+        self._ckpt_uri = ckpt_uri
+        self._eval_model_uri = eval_model_uri
+        self._session_options = session_options
+        if providers is None:
+            self._providers = C.get_available_providers()
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/engine/training_module.py:30]: We probably only need training mode and _model. All the other member variables do not need to be kept around (at least for now).

File: orttraining/orttraining/python/training/onnxblock/engine/training_module.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+        self._create_training_module()
+
+    def __call__(self, *input):
+
+        if self.training:
+            self._model.train_step(*input, self.fetches)
+        elif self.training == False:
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/engine/training_module.py:38]: you can just use `else` here.

File: orttraining/orttraining/python/training/onnxblock/engine/training_module.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        This method is responsible for creating the model and initializing the parameters.
+        """
+        # Create session options as a default value.
+        self._session_options = None
+
+        # Create session options as a default value.
+        self._env = None
+        # Module is not exposed yet
+        model = C.Module(self._train_model_uri, self._ckpt_uri, self._eval_model_uri)
+        self._model = model
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/engine/training_module.py:54]: combine to 1 line.

File: orttraining/orttraining/python/training/onnxblock/engine/training_module.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
+import numpy as np
+
+from onnxruntime.capi import _pybind_state as C
+from onnxruntime.capi.onnxruntime_pybind11_state import OrtValueVector
+
+
+class TrainingModule:
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/engine/training_module.py:7]: perhaps this can be called just `Module` since it will be imported from onnxruntime.training.

File: orttraining/orttraining/python/training/onnxblock/engine/training_optimizer.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
+from onnxruntime.capi import _pybind_state as C
+
+
+class TrainingOptimizer:
+    """
+    Class for running Optimize Step in Training.
+    This class is a wrapper of Optimizer Class.
+    """
+
+    def __init__(self, train_optimizer_uri, model, env=None, session_options=None, providers=None, **kwargs) -> None:
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/engine/training_optimizer.py:10]: we can keep the constructor simple and just keep it to a bare minimum. We can remove env, session_options, providers, kwargs

File: orttraining/orttraining/python/training/onnxblock/engine/training_module.py [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+        """
+        return self._model
+
+    def parameters(self):
+        """
+        Returns the parameters of the model.
+        """
+        return self._model.parameters()
+
+    def save_checkpoint(self, ckpt_uri):
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/engine/training_module.py:100]: For now, save checkpoint can be here. But we must move save and load checkpoint out of the Module class when possible.

Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/engine/training_module.py:100]: do we not have a load_checkpoint?

File: orttraining/orttraining/python/training/onnxblock/engine/training_optimizer.py [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+        # Optimizer is not exposed yet
+        optimizer = C.Optimizer(self._train_optimizer_uri, self._model)
+
+        self._optimizer = optimizer
+
+    def step(self):
+        """
+        Run Optimizer Step.
+        """
+        return self._optimizer.optimizer_step()
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/engine/training_optimizer.py:42]: what does this return?

File: orttraining/orttraining/python/training/onnxblock/engine/training_optimizer.py [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+        """
+        Saves the checkpoint.
+        """
+        return self._optimizer.save_checkpoint(ckpt_uri)
+
+    def load_checkpoint(self, ckpt_uri):
+        """
+        Loads the checkpoint.
+        """
+        return self._optimizer.load_checkpoint(ckpt_uri)
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/engine/training_optimizer.py:54]: How do these save and load checkpoint interact with the ones in TrainingModule?

File: orttraining/orttraining/python/training/onnxblock/engine/training_module.py (hunk 3 above)
Comment (AdamLouly) [reply orttraining/orttraining/python/training/onnxblock/engine/training_module.py:41]: So users do not have to create an empty ortVector and pass it, so we just initialize it in the object creation time.

File: orttraining/orttraining/python/training/onnxblock/engine/training_module.py (hunk 4 above)
Comment (AdamLouly) [reply orttraining/orttraining/python/training/onnxblock/engine/training_module.py:43]: why would we do that if its only being used for TrainingModule and no where else?

File: orttraining/orttraining/python/training/onnxblock/engine/training_module.py (hunk 6 above)
Comment (AdamLouly) [reply orttraining/orttraining/python/training/onnxblock/engine/training_module.py:92]: this returns the model to be passed for optimizer

File: orttraining/orttraining/python/training/onnxblock/engine/training_module.py (hunk 14 above)
Comment (AdamLouly) [reply orttraining/orttraining/python/training/onnxblock/engine/training_module.py:100]: no, Load checkpoint is basically calling the constructor, to create a new model with checkpoint url, there is no load checkpoint on c++ side

File: orttraining/orttraining/python/training/onnxblock/engine/training_module.py (hunk 2 above)
Comment (AdamLouly) [reply orttraining/orttraining/python/training/onnxblock/engine/training_module.py:34]: It accepts OrtValueVectors for now.

File: orttraining/orttraining/python/training/onnxblock/engine/training_optimizer.py (hunk 15 above)
Comment (AdamLouly) [reply orttraining/orttraining/python/training/onnxblock/engine/training_optimizer.py:42]: it returns None.

File: orttraining/orttraining/python/training/onnxblock/engine/training_module.py (hunk 7 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/onnxblock/engine/training_module.py:112]: Correcting myself. I think we can make this assumption, but probably should not.
In any case, i think we can let the users parse the output the way they want, and we don't need to provide this utility function to them.

File: orttraining/orttraining/python/training/onnxblock/engine/training_module.py (hunk 4 above)
Comment (AdamLouly) [reply orttraining/orttraining/python/training/onnxblock/engine/training_module.py:43]: I'll add a TODO for this for the future

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 17]
```diff
@@ -816,6 +816,50 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
...
 #ifdef ENABLE_TRAINING_ON_DEVICE
+  py::class_<onnxruntime::training::api::Module> training_module(m, "Module", R"pbdoc(Training Module Class.)pbdoc");
+  training_module.def(py::init([](const std::string model_uri, const std::string ckpt_uri, std::optional<std::string> eval_model_uri) {
+                   onnxruntime::training::api::CheckpointState state;
+                   auto checkpoint_to_load_path = ckpt_uri;
+                   ORT_THROW_IF_ERROR(onnxruntime::training::api::LoadCheckpoint(checkpoint_to_load_path, state));
+
+                   onnxruntime::SessionOptions session_option;
+                   return std::make_unique<onnxruntime::training::api::Module>(model_uri,
+                                                                               state.module_checkpoint_state.named_parameters, session_option,
```
Comment (askhade) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:827]: nit: fix all the cpplint warnings

File: orttraining/orttraining/python/training/onnxblock/engine/module.py [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+
+    def reset_grad(self):
+        """
+        Resets the gradient of the parameters.
+        """
+        return self._model.reset_grad()
+
+    def get_model(self):
+        """
+        Returns the model to be passed to initialize the Optimizer.
```
Comment (askhade) [inline orttraining/orttraining/python/training/onnxblock/engine/module.py:83]: May be change the comment to - Returns the module class to be passed to initialize the Optimizer?

File: orttraining/orttraining/python/training/onnxblock/engine/training_optimizer.py [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+        self._model = model.get_model()
+
+        self._create_training_optimizer()
+
+    def _create_training_optimizer(self):
+        """
+        This method is responsible for creating the optimizer and initializing the parameters.
+        """
+        # TODO : make this as a util function.
+        self._optimizer = C.Optimizer(self._train_optimizer_uri, self._model)
```
Comment (askhade) [inline orttraining/orttraining/python/training/onnxblock/engine/training_optimizer.py:24]: can this be part of the init mtd?

File: orttraining/orttraining/test/python/orttraining_test_python_bindings.py [hunk 20]
```diff
@@ -0,0 +1,243 @@
...
+        checkpoint_file_path = os.path.join(temp_dir, "checkpoint")
+        onnxblock.save_checkpoint((trainable_params, non_trainable_params), checkpoint_file_path)
+
+        model_file_path = os.path.join(temp_dir, "training_model.onnx")
+        onnx.save(onnx_model, model_file_path)
+
+        # Create a Module.
+        model = Module(model_file_path, checkpoint_file_path)
+        model.train()
+        fetches = model(forward_inputs)
```
Comment (askhade) [inline orttraining/orttraining/test/python/orttraining_test_python_bindings.py:136]: since you have the pytorch model can you compare the outputs?

File: orttraining/orttraining/test/python/orttraining_test_python_bindings.py [hunk 21]
```diff
@@ -0,0 +1,243 @@
...
+    # Initialize Models
+    simple_model, onnx_model, _, eval_model = _create_training_models()
+    trainable_params, non_trainable_params = simple_model.parameters()
+
+    # Generating random data for testing.
+    inputs = torch.randn(64, 784).numpy()
+    labels = torch.randint(high=10, size=(64,), dtype=torch.int32).numpy()
+    forward_inputs = OrtValueVector()
+    forward_inputs.reserve(2)
+    forward_inputs.push_back(OrtValue.ortvalue_from_numpy(inputs)._ortvalue)
```
Comment (askhade) [inline orttraining/orttraining/test/python/orttraining_test_python_bindings.py:150]: Conversion from numpy to OrtValueVector can be extracted into a util mtd. Can we add this util mtd in our product code to make it easier for the users to go from numpy to OrtValueVector?

File: orttraining/orttraining/training_api/include/module.h [hunk 22]
```diff
@@ -101,7 +101,7 @@ struct Module {
   std::vector<OrtValue> weights_;
   std::vector<OrtValue> gradients_;
   bool accumulate_gradient_ = false;
-  const std::unordered_map<std::string, std::shared_ptr<Parameter>>& named_parameters_;
+  const std::unordered_map<std::string, std::shared_ptr<Parameter>> named_parameters_;
```
Comment (askhade) [inline orttraining/orttraining/training_api/include/module.h:104]: Why is this change required?

File: orttraining/orttraining/test/python/orttraining_test_python_bindings.py (hunk 21 above)
Comment (AdamLouly) [reply orttraining/orttraining/test/python/orttraining_test_python_bindings.py:150]: added a TODO for the next PR.

File: orttraining/orttraining/training_api/include/module.h (hunk 22 above)
Comment (AdamLouly) [reply orttraining/orttraining/training_api/include/module.h:104]: Initializing the optimizer using the model's parameters did not work previously, needed that change to fix it.

File: orttraining/orttraining/test/python/orttraining_test_python_bindings.py (hunk 20 above)
Comment (AdamLouly) [reply orttraining/orttraining/test/python/orttraining_test_python_bindings.py:136]: to compare that, we have to expose the Parameter Object first, this will be included in the next PR.

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 23]
```diff
@@ -821,8 +822,52 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
...
+      .def("reset_grad", [](onnxruntime::training::api::Module* model) -> void {
+        ORT_THROW_IF_ERROR(model->ResetGrad());
+      })
+      .def("save_checkpoint", [](onnxruntime::training::api::Module* model, const std::string& checkpoint_path) -> void {
+        onnxruntime::training::api::CheckpointState state;
+        ORT_THROW_IF_ERROR(model->GetStateDict(state.module_checkpoint_state));
+        ORT_THROW_IF_ERROR(onnxruntime::training::api::SaveCheckpoint(state,
+                                                                      ToPathString(checkpoint_path)));
+      });
+  py::class_<onnxruntime::training::api::Optimizer> training_optimizer(m, "Optimizer", R"pbdoc(Training Optimizer Class.)pbdoc");
```
Comment (baijumeswani) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:856]: add a new line to denote another class?

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 24]
```diff
@@ -821,8 +822,52 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
...
+                                                                                     GetTrainingORTEnv(), std::vector<std::shared_ptr<IExecutionProvider>>());
+                    }))
+      .def("optimizer_step", [](onnxruntime::training::api::Optimizer* optimizer) -> void {
+        ORT_THROW_IF_ERROR(optimizer->Step());
+      });
+  py::class_<onnxruntime::training::api::Parameter> parameter(m, "Parameter", R"pbdoc(Parameter Class.)pbdoc");
+  parameter.def(py::init([](const std::string name, OrtValue& data, bool requires_grad) {
+    return std::make_unique<onnxruntime::training::api::Parameter>(name,
+                                                                   data, requires_grad);
+  }));
```
Comment (baijumeswani) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:870]: why is this needed?

File: orttraining/orttraining/python/training/onnxblock/__init__.py [hunk 25]
```diff
@@ -7,5 +7,6 @@
 from . import loss, optim
 from .building_blocks import Block
 from .checkpoint_utils import load_checkpoint_to_model, save_checkpoint
+from .engine import Module, Optimizer
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/__init__.py:10]: Why is the execution code defined inside `onnxblock`?

File: orttraining/orttraining/python/training/onnxblock/engine/__init__.py [hunk 26]
```diff
@@ -0,0 +1,2 @@
+from .module import Module
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/engine/__init__.py:1]: Can we come up with a more marketable name than engine?

File: orttraining/orttraining/python/training/onnxblock/engine/module.py [hunk 27]
```diff
@@ -0,0 +1,90 @@
+from onnxruntime.capi import _pybind_state as C
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/engine/module.py:1]: Please add the license at the beginning of the python file.

File: orttraining/orttraining/python/training/onnxblock/engine/module.py [hunk 28]
```diff
@@ -0,0 +1,90 @@
...
+        Initializes Model for Training.
+        __init__ will call an internatl function to create the model.
+        """
+        # TODO : Add support for bytes on train_model_uri and eval_model_uri.
+        self.training = True
+        self._train_model_uri = train_model_uri
+        self._ckpt_uri = ckpt_uri
+        self._eval_model_uri = eval_model_uri
+
+        self._create_training_module()
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/engine/module.py:24]: Since the paths are not needed after the construction, we can just use them transactionally.
```suggestion
        self._create_training_module(train_model_uri, ckpt_uri, eval_model_uri)
```

File: orttraining/orttraining/python/training/onnxblock/engine/module.py [hunk 29]
```diff
@@ -0,0 +1,90 @@
...
+
+    def __call__(self, *input):
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/engine/module.py:26]: python has built-in `input`. Can we make it more explicit by calling it `user_inputs` ?

File: orttraining/orttraining/python/training/onnxblock/engine/module.py [hunk 30]
```diff
@@ -0,0 +1,90 @@
...
+        """
+        Resets the gradient of the parameters.
+        """
+        return self._model.reset_grad()
+
+    def get_model(self):
+        """
+        Returns the module class to be passed to initialize the Optimizer.
+        """
+        return self._model
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/onnxblock/engine/module.py:83]: Is it possible to avoid exposing an internal type to the end user? Let's define `get_model` to be used only internally by renaming it to `_get_model`. Alternatively, you can get rid of the method altogether and have the optimizer access `model._model` as needed.

File: orttraining/orttraining/test/python/orttraining_on_device_training_tests.py [hunk 31]
```diff
@@ -24,6 +24,15 @@ def parse_arguments():
     return parser.parse_args()
 
 
+def run_on_device_training_api_tests(cwd, log):
```
Comment (baijumeswani) [inline orttraining/orttraining/test/python/orttraining_on_device_training_tests.py:27]:
```suggestion
def run_on_device_training_python_api_tests(cwd, log):
```

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 32]
```diff
@@ -821,8 +822,52 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
...
           GradientDefinitionRegistry::Instance().SetStopGradientEdgesForNode(key, edges);
         });
-
 #ifdef ENABLE_TRAINING_ON_DEVICE
+  py::class_<onnxruntime::training::api::Module> training_module(m, "Module", R"pbdoc(Training Module Class.)pbdoc");
+  training_module
+      .def(py::init([](const std::string model_uri, const std::string ckpt_uri, std::optional<std::string> eval_model_uri) {
+        onnxruntime::training::api::CheckpointState state;
+        auto checkpoint_to_load_path = ckpt_uri;
+        ORT_THROW_IF_ERROR(onnxruntime::training::api::LoadCheckpoint(checkpoint_to_load_path, state));
```
[the remark this answers is not in the document]
Comment (askhade) [reply orttraining/orttraining/python/orttraining_pybind_state.cc:831]: +1 for holding onto the state here

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 33]
```diff
@@ -821,8 +822,52 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
...
+
+        onnxruntime::SessionOptions session_option;
+        return std::make_unique<onnxruntime::training::api::Module>(model_uri,
+                                                                    state.module_checkpoint_state.named_parameters, session_option,
+                                                                    GetTrainingORTEnv(), std::vector<std::shared_ptr<IExecutionProvider>>(), eval_model_uri);
```
[the remark this answers is not in the document]
Comment (askhade) [reply orttraining/orttraining/python/orttraining_pybind_state.cc:836]: Can you also add a TODO to add support for CUDA

File: orttraining/orttraining/python/training/onnxblock/engine/__init__.py (hunk 26 above)
Comment (askhade) [reply orttraining/orttraining/python/training/onnxblock/engine/__init__.py:1]: +1

File: orttraining/orttraining/python/training/onnxblock/engine/module.py (hunk 28 above)
Comment (askhade) [reply orttraining/orttraining/python/training/onnxblock/engine/module.py:24]: another approach is to get rid of _create_training_module since it does not really do much

File: orttraining/orttraining/test/python/orttraining_test_python_bindings.py [hunk 34]
```diff
@@ -0,0 +1,244 @@
...
+    simple_model, onnx_model, _, _ = _create_training_models()
+    trainable_params, non_trainable_params = simple_model.parameters()
+
+    # Generating random data for testing.
+    inputs = torch.randn(64, 784).numpy()
+    labels = torch.randint(high=10, size=(64,), dtype=torch.int32).numpy()
+    forward_inputs = OrtValueVector()
+    forward_inputs.reserve(2)
+    forward_inputs.push_back(OrtValue.ortvalue_from_numpy(inputs)._ortvalue)
+    forward_inputs.push_back(OrtValue.ortvalue_from_numpy(labels)._ortvalue)
```
Comment (baijumeswani) [inline orttraining/orttraining/test/python/orttraining_test_python_bindings.py:123]: are users expected to pass in `_ortvalue`? Can this be handled internally by our function? Users should only need to be aware of the python `OrtValue`.

File: orttraining/orttraining/test/python/orttraining_test_python_bindings.py (hunk 21 above)
Comment (baijumeswani) [reply orttraining/orttraining/test/python/orttraining_test_python_bindings.py:151]: Do we need an OrtValueVector? Can we not use a python list of OrtValues? Let's tr to avoid any non-pythonic structs in the user interface if possible.

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 35]
```diff
@@ -821,8 +822,52 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
         [](const std::string& key, const std::unordered_set<size_t> edges) -> void {
           GradientDefinitionRegistry::Instance().SetStopGradientEdgesForNode(key, edges);
         });
-
 #ifdef ENABLE_TRAINING_ON_DEVICE
+  py::class_<onnxruntime::training::api::Module> training_module(m, "Module", R"pbdoc(Training Module Class.)pbdoc");
+  training_module
+      .def(py::init([](const std::string model_uri, const std::string ckpt_uri, std::optional<std::string> eval_model_uri) {
```
[the remark this answers is not in the document]
Comment (AdamLouly) [reply orttraining/orttraining/python/orttraining_pybind_state.cc:828]: in python its mapped to a default None value : 
`  def __init__(self, train_model_uri, ckpt_uri, eval_model_uri=None) -> None:`

File: orttraining/orttraining/python/orttraining_pybind_state.cc (hunk 24 above)
Comment (AdamLouly) [reply orttraining/orttraining/python/orttraining_pybind_state.cc:870]: its not needed for now, will delete

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 36]
```diff
@@ -816,6 +816,36 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
         });
 
 #ifdef ENABLE_TRAINING_ON_DEVICE
+  py::class_<onnxruntime::training::api::Module> training_module(m, "Module", R"pbdoc(Training Module Class.)pbdoc");
+  training_module.def(py::init([](const std::string model_uri, const std::string ckpt_uri) {
+                   onnxruntime::training::api::CheckpointState state;
+                   auto checkpoint_to_load_path = ckpt_uri;
```
Comment (askhade) [inline orttraining/orttraining/python/orttraining_pybind_state.cc]: what is the purpose of this?

File: orttraining/orttraining/test/python/orttraining_test_python_bindings.py [hunk 37]
```diff
@@ -0,0 +1,244 @@
...
+
+    with tempfile.TemporaryDirectory() as temp_dir:
+        # Save models & checkpoint files to load them later.
+        checkpoint_file_path = os.path.join(temp_dir, "checkpoint")
+        onnxblock.save_checkpoint((trainable_params, non_trainable_params), checkpoint_file_path)
```
Comment (askhade) [inline orttraining/orttraining/test/python/orttraining_test_python_bindings.py:191]: there is a lot of duplicate code in all the tests. Can you simply this.

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 38]
```diff
@@ -821,8 +822,52 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
         [](const std::string& key, const std::unordered_set<size_t> edges) -> void {
           GradientDefinitionRegistry::Instance().SetStopGradientEdgesForNode(key, edges);
         });
-
 #ifdef ENABLE_TRAINING_ON_DEVICE
+  py::class_<onnxruntime::training::api::Module> training_module(m, "Module", R"pbdoc(Training Module Class.)pbdoc");
```
Comment (askhade) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:826]: Training Module

File: orttraining/orttraining/python/training/onnxblock/engine/__init__.py (hunk 26 above)
Comment (AdamLouly) [reply orttraining/orttraining/python/training/onnxblock/engine/__init__.py:1]: any suggestions?

File: orttraining/orttraining/python/training/onnxblock/__init__.py (hunk 25 above)
Comment (AdamLouly) [reply orttraining/orttraining/python/training/onnxblock/__init__.py:10]: so we can call them directly like that : from onnxruntime.training.onnxblock import Module, Optimizer

File: orttraining/orttraining/test/python/orttraining_test_python_bindings.py [hunk 39]
```diff
@@ -0,0 +1,159 @@
+import os
+import tempfile
+
+import onnx
+import torch
+from orttraining_test_onnxblock import _get_models
+
+import onnxruntime.training.onnxblock as onnxblock
+from onnxruntime.training.onnxblock.engine import Module, Optimizer
```
Comment (askhade) [inline orttraining/orttraining/test/python/orttraining_test_python_bindings.py:9]: should this be from onnxruntime.training.engine import Module, Optimizer

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 40]
```diff
@@ -828,23 +828,24 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
   py::class_<onnxruntime::training::api::Module> training_module(m, "Module", R"pbdoc(Training Module.)pbdoc");
   training_module
       .def(py::init([](const std::string model_uri,
-                       const std::string& ckpt_uri,
+                       onnxruntime::training::api::CheckpointState state,
```
Comment (askhade) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:831]: take a reference?

File: orttraining/orttraining/python/training/api/module.py [hunk 41]
```diff
@@ -0,0 +1,83 @@
...
+        self.training = True
+        self._model = C.Module(train_model_uri, state._state, eval_model_uri)
+
+    def __call__(self, user_inputs):
+        """
+        This method enables calling Module as a function to run the model.
+        Args:
+            user_inputs (OrtValueVector): input vector of ortvalues.
+        Returns:
+            fetches : OrtValueVector that has model output.
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/api/module.py:33]: Can user inputs and fetches both be python lists?

File: orttraining/orttraining/test/python/orttraining_test_python_bindings.py (hunk 20 above)
Comment (baijumeswani) [reply orttraining/orttraining/test/python/orttraining_test_python_bindings.py:82]: Sorry, I may have missed this. Why do you need the Parameters to compare the output values?

Comment (AdamLouly) [reply orttraining/orttraining/test/python/orttraining_test_python_bindings.py:82]: so we can pull the parameters object and compare before and after to make sure they got changed after running the train, and optimizer step? 
or do you want to compare just the loss to make sure it got improved after a train, optimize step?

Comment (baijumeswani) [reply orttraining/orttraining/test/python/orttraining_test_python_bindings.py:82]: The outputs of the `TrainStep` include only the forward outputs. In this case, `fetches` would be the forward outputs. And should be the same as pytorch's forward outputs. And therefore, can be compared.

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 42]
```diff
@@ -828,23 +828,24 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
   py::class_<onnxruntime::training::api::Module> training_module(m, "Module", R"pbdoc(Training Module.)pbdoc");
   training_module
       .def(py::init([](const std::string model_uri,
```
Comment (askhade) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:830]: const std:string& ?

File: orttraining/orttraining/training_api/include/module.h [hunk 43]
```diff
@@ -111,7 +111,7 @@ struct Module {
...
```
Comment (askhade) [inline orttraining/orttraining/training_api/include/module.h:114]: Now that you are exposing Checkpoint State and it outlives Module class, do we need this change?

File: orttraining/orttraining/python/training/api/module.py (hunk 41 above)
Comment (askhade) [reply orttraining/orttraining/python/training/api/module.py:33]: I agree! Conversion to OrtValue can happen internally

**Review by baijumeswani — DISMISSED** [review]

**Review by baijumeswani — APPROVED** [review]

**Review by askhade — APPROVED** [review]

**Comment by AdamLouly** [comment]
Thanks for the reivew
